designs/sam-config.md

```diff
@@ -0,0 +1,413 @@
+
+Future
+----------
+
+In the future, based on the file names of the configuration files, the environment could also be inferred.
+
+```
+.aws-sam/samconfig-dev.toml
+.aws-sam/samconfig-beta.toml
+.aws-sam/samconfig-prod.toml
+```
+
+`--env` dev will refer to `.aws-sam/samconfig-dev.toml` and so on.
+
+> I'm not in favor of parsing file names in lieu of TOML table/section prefixes. What are the TOML tables in the first file? Does it contain `[dev.build.parameters]` or just `[build.parameters]` or `[default.build.parameters]`. First is redundant, second is out-of-spec, third is potentially confusing since there could be defaults elsewhere.
+> Instead, I support loading of multiple config files in order they are specified. Similar to how `docker-compose` can have multiple `-f filename` parameters. However, then comes the discussion of merge rules: overwrite/append keys, extend arrays, and application of these merge rules across files across defaults, merge rules for config files of different spec versions, etc.?
+
+If multiple default file locations are added in the look up order for `samconfig.toml`, this means that multiple config files can be merged together.
+
+For example, if the hierachy of lookup for configuration files are: $SAM_CLI_CONFIG -> `.aws-sam/samconfig.toml` -> `~/.aws-sam/samconfig.toml`
+
+The resulting configuration would be a merge of all the sections that are relevant for the command that was run.
+
+This way, configuration that might be global can be placed in `~/.aws-sam/samconfig.toml`.
+
+```
+version = 0.1
+[default.build.parameters]
+use_container = True
+skip_pull_image = True
+```
+
+Project specific configuration placed in `.aws-sam/samconfig.toml`
+
+```
+version = 0.1
+[default.build.parameters]
+parameter_overrides="ParameterKey=KeyPairName,ParameterValue=MyKey ParameterKey=InstanceType,ParameterValue=t1.micro"
+```
+
+Eventual merged configuration read during `sam build` in-memory.
+
+```
+version = 0.1
+[default.build.parameters]
+use_container = True
+skip_pull_image = True
+parameter_overrides="ParameterKey=KeyPairName,ParameterValue=MyKey ParameterKey=InstanceType,ParameterValue=t1.micro"
+```
+
+Open Questions
+-------------------------------
+
+* Potentially every sam command could have functionality to have a series of command line parameters exported into a configuraion file.
+
+
+Out-of-Scope
+------------
+
+* Not focusing on a global configuration. SAM CLI already has a notion of a global config at `~/.aws-sam/metadata.json`
+
+> This is confusing/conflicting with a few paragraphs above is written, "This way, configuration that might be global can be placed...". While technically such could be possible, I recommend not writing of a technically possible scenario that conflicts with what is stated as out-of-scope.
```

Comment:
I think this was the effect of the back and forth on the design, will take this into consideration with the refactor on the design doc to accurately reflect present day and paths forward.